### ISSUE TITLE
allow zero-valent tuple type

### ISSUE THREAD

**Issue by aleksey-bykov** [Microsoft/TypeScript#13239]
the `{}` type is allowed, while `[]` is not allowed, which is a sad news because i was planning to use it as a unit type distinct from `null` and `undefined` which have their own meaning already

consider allowing the empty tuple type

**Comment by aleksey-bykov** [comment]
the `length` property that is always `0`

**Comment by aleksey-bykov** [comment]
@chicoxyzzy `[]` to `[a]` is what `0` is to `1`, an identity type if you will: `[] + [a] = [a]`,  `[a] + [] = [a]`

https://homotopytypetheory.org/2011/04/18/whats-special-about-identity-types/

still confused?

**Comment by chicoxyzzy** [comment]
@aleksey-bykov TBH I don't understand the use case for such unit type.

**Comment by chicoxyzzy** [comment]
also `{}` is not unit type and it behaves similarly `[]`

**Comment by aleksey-bykov** [comment]
>  TBH I don't understand the use case for such unit type.

as i said it could have been yet another unit type in addition to `null` and `undefined`

> also {} is not unit type and it behaves similarly []

`{}` is a valid type in TS and `[]` is gives an error, not sure what you are talking about

```
const values : [] = [];
```

![image](https://cloud.githubusercontent.com/assets/937933/21584137/2c5fb2bc-d06a-11e6-9f70-1cbd5cb3e3f3.png)

**Comment by chicoxyzzy** [comment]
If `[]` will ever exist it should be an alias to `any[]` similarly to how `{}` is an alias to `{ [propName: string]: any }`

**Comment by aleksey-bykov** [comment]
`any[]` implies you can add elements to it:

```
const values: any[] = [];
values.push(1);
```

whereas `[]` is meant to be empty, that's the difference

```
const values: []= []; // hypothetical situation
values.push(1); // `push` method is not defined on `[]`
```

**Comment by chicoxyzzy** [comment]
In that way `[]` will behave differently to `{}` which doesn't hide any object prototype methods so that is counterintuitive

**Comment by aleksey-bykov** [comment]
well it was never said that `[]` should behave anything like `{}` in the first place, not sure why we are trying to compare them

> {} which doesn't hide any object prototype methods so that is counterintuitive

does official TS guideline looks counterintuitive to you? yet you can find `ReadOnlyArray` there which hides its prototype methods, riddle me that:

https://www.typescriptlang.org/docs/handbook/interfaces.html


![image](https://cloud.githubusercontent.com/assets/937933/21584236/fa26aea4-d06e-11e6-85eb-efe0f8444a11.png)

**Comment by chicoxyzzy** [comment]
The counterintuitive part is that `[]` and `{}` behaviour will be different in your proposal. Readonly types indicate their behaviour explicitly and that's ok.

**Comment by chicoxyzzy** [comment]
Also as you mentioned you can declare your array unit
```ts
const arrayUnit: ReadonlyArray<any> = [];
```

**Comment by aleksey-bykov** [comment]
again, i don't care what `{}` is, it's outside of this topic (as far as i can tell), whatever behaviour it has it should be kept intact

here i was trying to talk about a different type which i called `[]`, and the proposition is for this type states that it possesses some unique properties yet very similar to `ReadOnlyArray` (except for the index signature) which has been around since #cac3a32513d2f06cbf1d7da0e34b4df571fe472b

> Also as you mentioned you can declare your array unit

no, i cannot declare it like that because i need it to be nonassignable from `ReadOnlyArray<any>`

**Comment by aleksey-bykov** [comment]
and this is why:
```
const values: ReadOnlyArray<number> = [1, 2, 3];
let empty: [] = [];
empty = values; //<-- if ReadOnlyArray<a> is assignable, then the whole idea of having a type for an  EMPTY tuple doesn't make sense
```

**Comment by chicoxyzzy** [comment]
```ts
const empty: ReadonlyArray<any> = [];

empty.push(0); // error
empty = [1, 2, 3]; // error
```

**Comment by aleksey-bykov** [comment]
```
const empty: ReadonlyArray<any> = [1, 2, 3];

empty.push(0); // error, but who cares it already has [1, 2, 3] instead of being empty
empty = [1, 2, 3]; // error, again too late
```

**Comment by chicoxyzzy** [comment]
This is strange when you assign non-empty array to a constant variable named `empty`. Can you provide an example?

**Comment by aleksey-bykov** [comment]
closed in favor of #13126
